### DISCUSSION TITLE
How to add EFI variable storage to a LoongArch64 VM?

### DISCUSSION THREAD

**Discussion by asdfugil** [utmapp/UTM#6369] [UTM Help (macOS)]
Host information:
OS: macOS 14.4.1
UTM: UTM 4.5.2
Chip: Apple M1 Pro

I cannot figure out how to add EFI variable storage to a LoongArch64 VM, so I cannot properly install grub in the VM.
It *does* boot fine if I hack it and copy `/boot/efi/EFI/debian/grubloongarch64.efi` to `/boot/efi/EFI/BOOT/BOOTLOONGARCH64.EFI`

config plist: [config.plist.zip](https://github.com/utmapp/UTM/files/15292237/config.plist.zip)
UEFI firmware: [QEMU_EFI_7.2.fd.zip](https://github.com/utmapp/UTM/files/15292294/QEMU_EFI_7.2.fd.zip)

**Comment by osy** [comment]
You can do it from the EFI Shell or using an .efi application (for example, copy it to an .iso and mount it and run EFI Shell).

**Reply by asdfugil** [reply]
The problem is there is no pflash for the EFI variable storage and I have not figured out how to add it in UTM so whatever done to efi variables in the VM will not be saved.

**Reply by osy** [reply]
There's two ways to do this. You can manually add your pflash for both EFI ROM and EFI variables by adding "drives" and setting the interface to PFlash. Or you can open up the .utm bundle and manually modify the efi_vars.fd file. But note that it is a qcow2 format so you would need to convert the raw efi_vars.fd to qcow2.

**Reply by asdfugil** [reply]
Trying to add pflash as you described:
```
QEMU error: QEMU exited from an error: qemu-loongarch64-softmmu: -drive if=pflash,media=disk,id=drive66476493-99BA-4539-84C6-4396CB736258,file=/Users/nick/Downloads/Debian loong64.utm/Data/QEMU_VARS.fd,discard=unmap,detect-zeroes=unmap: machine type does not support if=pflash,bus=0,unit=0
```
I think, the QEMU version shipped in UTM may not support LoongArch pflash, the machine type is only added in QEMU 7.1
I can find references to QEMU LoongArch pflash online though, none of that worked in UTM.

**Reply by osy** [reply]
We're on 7.2.0 so it should work... If you can find the right command to get it working on QEMU, open a new issue and I'll try to add that into the GUI in the future.
